# MVP Predictions in NBA with Machine Learning

We've scraped the data on `1-web_scraping.ipynb`, clean them up in `2-data_cleaning.ipynb`, now we'll apply machine learning to predict the MVP based on players' stats we've collected. We'll be experimenting with several models below. Let's go, shall we?

Most machine learning models need to have full data--it won't work when there are no missing values. So we get to check that beforehand.

In [1]:
import pandas as pd

stats = pd.read_csv('player_mvp_stats.csv')

In [2]:
stats

Unnamed: 0          Player Pos  Age   Tm   G  GS    MP   FG   FGA  ...  \
0               0      A.C. Green  PF   36  LAL  82  82  23.5  2.1   4.7  ...   
1               1      Brian Shaw  SG   33  LAL  74   2  16.9  1.7   4.4  ...   
2               2    Derek Fisher  PG   25  LAL  78  22  23.1  2.1   6.2  ...   
3               3   Devean George  SF   22  LAL  49   1   7.0  1.1   2.9  ...   
4               4       Glen Rice  SF   32  LAL  80  80  31.6  5.3  12.3  ...   
...           ...             ...  ..  ...  ...  ..  ..   ...  ...   ...  ...   
11514       11514   Spencer Hawes  PF   28  MIL  54   1  14.8  2.5   5.1  ...   
11515       11515     Steve Novak  PF   33  MIL   8   0   2.8  0.3   0.9  ...   
11516       11516  Terrence Jones  PF   25  MIL  54  12  23.5  4.3   9.1  ...   
11517       11517      Thon Maker   C   19  MIL  57  34   9.9  1.5   3.2  ...   
11518       11518      Tony Snell  SG   25  MIL  80  80  29.2  3.1   6.8  ...   

       Pts Max  Share                Team   W   L   W/L%   GB   PS/G   PA/G  \
0          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
1          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
2          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
3          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
4          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
...        ...    ...                 ...  ..  ..    ...  ...    ...    ...   
11514      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   
11515      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   
11516      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   
11517      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   
11518      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   

        SRS  
0      8.41  
1      8.41  
2      8.41  
3      8.41  
4      8.41  
...     ...  
11514 -0.45  
11515 -0.45  
11516 -0.45  
11517 -0.45  
11518 -0.45  

[11519 rows x 42 columns]

In [3]:
del stats['Unnamed: 0']

In [4]:
stats.isna().sum()

Player        0
Pos           0
Age           0
Tm            0
G             0
GS            0
MP            0
FG            0
FGA           0
FG%          45
3P            0
3PA           0
3P%        1484
2P            0
2PA           0
2P%          87
eFG%         45
FT            0
FTA           0
FT%         446
ORB           0
DRB           0
TRB           0
AST           0
STL           0
BLK           0
TOV           0
PF            0
PTS           0
Year          0
Pts Won       0
Pts Max       0
Share         0
Team          0
W             0
L             0
W/L%          0
GB            0
PS/G          0
PA/G          0
SRS           0
dtype: int64

Missing values all consists of percentages of plays. This implies that there are players who did not participate in `FG`, `3P`, `2P`, and `FT`.

In [5]:
stats[stats['FG%'].isna()][['Player', 'FG', 'FGA', 'FG%']]

Player   FG  FGA  FG%
105             Guy Rucker  0.0  0.0  NaN
308             Gani Lawal  0.0  0.0  NaN
995             C.J. Miles  0.0  0.0  NaN
1474            Ade Murkey  0.0  0.0  NaN
1694          Ronny Turiaf  0.0  0.0  NaN
1869         DeJon Jarreau  0.0  0.0  NaN
1908           Lari Ketner  0.0  0.0  NaN
2447             Ben Moore  0.0  0.0  NaN
2462   Trey McKinney-Jones  0.0  0.0  NaN
2563        Josh Childress  0.0  0.0  NaN
2740            Andy Panko  0.0  0.0  NaN
2759       Dickey Simpkins  0.0  0.0  NaN
2821       Pavel Podkolzin  0.0  0.0  NaN
2976      Jerrelle Benimon  0.0  0.0  NaN
3086            Tyler Hall  0.0  0.0  NaN
3270           Jordan Bell  0.0  0.0  NaN
3456           Alex Scales  0.0  0.0  NaN
3705           J.P. Macura  0.0  0.0  NaN
3710        Marques Bolden  0.0  0.0  NaN
3927              Deng Gai  0.0  0.0  NaN
4028      Alondes Williams  0.0  0.0  NaN
4424          Dwayne Jones  0.0  0.0  NaN
4489           Chris Smith  0.0  0.0  NaN
4957          Damion James  0.0  0.0  NaN
5464         Bryon Russell  0.0  0.0  NaN
5897         Brian Skinner  0.0  0.0  NaN
6108            Sam Dekker  0.0  0.0  NaN
6571          John Holland  0.0  0.0  NaN
6574          Kobi Simmons  0.0  0.0  NaN
6754         Ashton Hagans  0.0  0.0  NaN
6813       Jaime Echenique  0.0  0.0  NaN
7458      Hassan Whiteside  0.0  0.0  NaN
7515            Tyler Ulis  0.0  0.0  NaN
7767            Sam Dekker  0.0  0.0  NaN
7802         Scottie Lewis  0.0  0.0  NaN
8388         JamesOn Curry  0.0  0.0  NaN
8409          Matt Carroll  0.0  0.0  NaN
8973       Javin DeLaurier  0.0  0.0  NaN
9012        Olden Polynice  0.0  0.0  NaN
9120      Danuel House Jr.  0.0  0.0  NaN
10168      Randell Jackson  0.0  0.0  NaN
10423         Jelani McCoy  0.0  0.0  NaN
11259   Michael Foster Jr.  0.0  0.0  NaN
11312          Tyler Lydon  0.0  0.0  NaN
11366          George King  0.0  0.0  NaN

In [6]:
stats[stats['3P%'].isna()][['Player', '3P', '3PA', '3P%']]

Player   3P  3PA  3P%
6           John Salley  0.0  0.0  NaN
12        Travis Knight  0.0  0.0  NaN
17        Anthony Mason  0.0  0.0  NaN
23       Duane Causwell  0.0  0.0  NaN
29          Todd Fuller  0.0  0.0  NaN
...                 ...  ...  ...  ...
11488    Evan Eschmeyer  0.0  0.0  NaN
11489  Gheorghe Mureșan  0.0  0.0  NaN
11491     Jim McIlvaine  0.0  0.0  NaN
11497  Mark Hendrickson  0.0  0.0  NaN
11498      Michael Cage  0.0  0.0  NaN

[1484 rows x 4 columns]

In [7]:
stats[stats['2P%'].isna()][['Player', '2P', '2PA', '2P%']]

Player   2P  2PA  2P%
105            Guy Rucker  0.0  0.0  NaN
308            Gani Lawal  0.0  0.0  NaN
396         Anthony Brown  0.0  0.0  NaN
715        Josh McRoberts  0.0  0.0  NaN
995            C.J. Miles  0.0  0.0  NaN
...                   ...  ...  ...  ...
11176     Jamaal Franklin  0.0  0.0  NaN
11259  Michael Foster Jr.  0.0  0.0  NaN
11312         Tyler Lydon  0.0  0.0  NaN
11366         George King  0.0  0.0  NaN
11438        Trevor Keels  0.0  0.0  NaN

[87 rows x 4 columns]

In [8]:
stats[stats['FT%'].isna()][['Player', 'FT', 'FTA', 'FT%']]

Player   FT  FTA  FT%
26      Jamal Robinson  0.0  0.0  NaN
30       A.J. Bramlett  0.0  0.0  NaN
33     Benoit Benjamin  0.0  0.0  NaN
95         A.J. Guyton  0.0  0.0  NaN
105         Guy Rucker  0.0  0.0  NaN
...                ...  ...  ...  ...
11366      George King  0.0  0.0  NaN
11438     Trevor Keels  0.0  0.0  NaN
11446      Luke Zeller  0.0  0.0  NaN
11481      Malcolm Lee  0.0  0.0  NaN
11515      Steve Novak  0.0  0.0  NaN

[446 rows x 4 columns]

Technically given the calculations this would be correct; dividing FT by FTA of 0 results in missing values. But since these values mean that they don't participate, for that reason we'll fill the missing values with `0` instead.

In [9]:
stats = stats.fillna(0)

In [10]:
stats.isna().sum()

Player     0
Pos        0
Age        0
Tm         0
G          0
GS         0
MP         0
FG         0
FGA        0
FG%        0
3P         0
3PA        0
3P%        0
2P         0
2PA        0
2P%        0
eFG%       0
FT         0
FTA        0
FT%        0
ORB        0
DRB        0
TRB        0
AST        0
STL        0
BLK        0
TOV        0
PF         0
PTS        0
Year       0
Pts Won    0
Pts Max    0
Share      0
Team       0
W          0
L          0
W/L%       0
GB         0
PS/G       0
PA/G       0
SRS        0
dtype: int64

## Training The Machine Learning Model

Now we can proceed to training the data to a machine learning model. We'll be grouping the data into two: (1) all predictors consisting of numerical values for the machine learning (we'll get rid of non-numerical ones), and (2) what we're trying to predict--in this is case it's `Share`.

In [11]:
stats

Player Pos  Age   Tm   G  GS    MP   FG   FGA    FG%  ...  \
0          A.C. Green  PF   36  LAL  82  82  23.5  2.1   4.7  0.447  ...   
1          Brian Shaw  SG   33  LAL  74   2  16.9  1.7   4.4  0.382  ...   
2        Derek Fisher  PG   25  LAL  78  22  23.1  2.1   6.2  0.346  ...   
3       Devean George  SF   22  LAL  49   1   7.0  1.1   2.9  0.389  ...   
4           Glen Rice  SF   32  LAL  80  80  31.6  5.3  12.3  0.430  ...   
...               ...  ..  ...  ...  ..  ..   ...  ...   ...    ...  ...   
11514   Spencer Hawes  PF   28  MIL  54   1  14.8  2.5   5.1  0.484  ...   
11515     Steve Novak  PF   33  MIL   8   0   2.8  0.3   0.9  0.286  ...   
11516  Terrence Jones  PF   25  MIL  54  12  23.5  4.3   9.1  0.470  ...   
11517      Thon Maker   C   19  MIL  57  34   9.9  1.5   3.2  0.459  ...   
11518      Tony Snell  SG   25  MIL  80  80  29.2  3.1   6.8  0.455  ...   

       Pts Max  Share                Team   W   L   W/L%   GB   PS/G   PA/G  \
0          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
1          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
2          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
3          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
4          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
...        ...    ...                 ...  ..  ..    ...  ...    ...    ...   
11514      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   
11515      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   
11516      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   
11517      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   
11518      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   

        SRS  
0      8.41  
1      8.41  
2      8.41  
3      8.41  
4      8.41  
...     ...  
11514 -0.45  
11515 -0.45  
11516 -0.45  
11517 -0.45  
11518 -0.45  

[11519 rows x 41 columns]

In [12]:
stats.columns

Index(['Player', 'Pos', 'Age', 'Tm', 'G', 'GS', 'MP', 'FG', 'FGA', 'FG%', '3P',
       '3PA', '3P%', '2P', '2PA', '2P%', 'eFG%', 'FT', 'FTA', 'FT%', 'ORB',
       'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS', 'Year',
       'Pts Won', 'Pts Max', 'Share', 'Team', 'W', 'L', 'W/L%', 'GB', 'PS/G',
       'PA/G', 'SRS'],
      dtype='object')

In [13]:
predictors = [
       'Age', 'G', 'GS', 'MP', 'FG', 'FGA', 'FG%', '3P', '3PA', '3P%', 
       '2P', '2PA', '2P%', 'eFG%', 'FT', 'FTA', 'FT%', 'ORB',
       'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS', 'Year',
       'W', 'L', 'W/L%', 'GB', 'PS/G', 'PA/G', 'SRS']

We have a data of MVPs and its stats from `2000-2023`. In this project, we'll use all `22` years of data history to predict MVP in the recent season, which is in `2023`. We'll include the previous years in a variable of `train`, and `2023` in `test`.

In [14]:
train  = stats[ stats['Year']  < 2023]
test   = stats[ stats['Year'] == 2023]

Now we can start going into the machine learning areas. We'll go into the first one while figuring out the best configuration for our prediction.

### (1) [`Ridge`](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.RidgeCV.html#sklearn.linear_model.RidgeCV) : Ridge Regression

First model we're using is Ridge Regression. It's basically a Linear Regression but with some regularization, meaning it has an algorithm to prevent your model from overfitting, meaning that the model works well on the training set but not in test set where the model matters at.

In [15]:
from sklearn.linear_model import Ridge

reg = Ridge(alpha=0.1)

reg.fit(train[predictors], train['Share'])

predictions = reg.predict(test[predictors])
predictions = pd.DataFrame(predictions, columns=['predictions'],
                                         index = test.index)

In [16]:
predictions

predictions
48        0.003463
49        0.029130
50        0.036673
51        0.219932
52       -0.000393
...            ...
11434    -0.014089
11435     0.003881
11436    -0.011316
11437     0.018509
11438     0.010729

[539 rows x 1 columns]

In [17]:
combination = pd.concat([test[['Player', 'Share']], predictions], axis=1)

In [18]:
combination

Player  Share  predictions
48                A.J. Green  0.000     0.003463
49              Bobby Portis  0.000     0.029130
50               Brook Lopez  0.000     0.036673
51     Giannis Antetokounmpo  0.606     0.219932
52              Goran Dragić  0.000    -0.000393
...                      ...    ...          ...
11434      Mitchell Robinson  0.000    -0.014089
11435             Obi Toppin  0.000     0.003881
11436         Quentin Grimes  0.000    -0.011316
11437             RJ Barrett  0.000     0.018509
11438           Trevor Keels  0.000     0.010729

[539 rows x 3 columns]

Usually when we're dealing with regressions, we can evaluate the model by looking for errors, commonly `mean_squared_error`. But since we're dealing with a ***rank*** of MVPs **and also** we have mostly `Share` values of `0`, we need to pinpoint what metrics are proper to evaluate it. Here we'll try to make the rank for both in 2023 and predicted by the model.

In [19]:
## Making Rank for 2023 and predicted Ranks
combination = combination.sort_values('Share', ascending=False)
combination['Rk'] = list(range(1, combination.shape[0]+1))

combination = combination.sort_values('predictions', ascending=False)
combination['Rk_predicted'] = list(range(1, combination.shape[0]+1))

In [20]:
combination

Player  Share  predictions   Rk  Rk_predicted
51     Giannis Antetokounmpo  0.606     0.219932    3             1
143              Luka Dončić  0.010     0.194485    8             2
11256            Joel Embiid  0.915     0.188030    1             3
559             Nikola Jokić  0.674     0.176815    2             4
6433          Damian Lillard  0.000     0.143781   90             5
...                      ...    ...          ...  ...           ...
131              Chris Silva  0.000    -0.040469  356           535
9508           Jacob Gilyard  0.000    -0.042709  209           536
6441         Justise Winslow  0.000    -0.046045   82           537
6440           Justin Minaya  0.000    -0.046267   83           538
11261            P.J. Tucker  0.000    -0.050242  194           539

[539 rows x 5 columns]

Now we have a list of ranks in 2023 (`Rk`) and ranks predicted by the model (`Rk_pred_ridge`). One metrics to evaluate the model properly is with **Average Precision**. Here's how it works:
* We'll showcase the top 5 MVPs for both `actual` ranks and `predicted` ranks, 
    * When the `predicted` rank is in the `actual` top 5, it'll have a perfect score of `1`.
    * Otherwise, we'll look for the `predicted` ranks, and we'll define the score based on how long the rank is `seen`. When `found`, we'll divide it by the duration it's `seen`. The longer the search, the lower the score is.

In [21]:
def find_avg_precision(combination):
    
    actual = combination.sort_values('Share', ascending=False).head(5)
    predicted = combination.sort_values('predictions', ascending=False)
    
    ps = []
    found = 0
    seen = 1
    
    for index, row in predicted.iterrows():
        if row['Player'] in actual['Player'].values:
            found += 1
            ps.append(found/seen)
        seen += 1
    
    return (sum(ps) / len(ps))

In [22]:
find_avg_precision(combination)

0.7416666666666666

### Backtesting for Prediction in Every Year

To evaluate the model in the 23-whole-year records of dataset, we can do backtesting. In short it's doing the same thing but starting from 5th year onwards; so in this case `2005`, and then `2006` and so on until `2023`. The average score will be put as the accuracy of the model.

In [23]:
years = list(range(2000, 2024))

avg_precisions = []
all_predictions = []

## We need 5 years of data for predictions:
for year in years[5:] :
    train = stats[ stats['Year']  < year]
    test  = stats[ stats['Year'] == year]

    reg.fit(train[predictors], train['Share'])

    predictions = reg.predict(test[predictors])
    predictions = pd.DataFrame(predictions, columns=['predictions'],
                                              index=test.index)
    combination = pd.concat([test[['Player', 'Share']], predictions], axis=1)

    all_predictions.append(combination)
    avg_precisions.append(find_avg_precision(combination))    

We have a `Ridge` accuracy of 69.2577% for the prediction in all years.

Here we've evaluated the model using a parameter of ranks. For us to make it easier to evaluate other models, let's make a function to `add_ranks`. This will also make us easier to diagnose the model in predicting the ranks.

In [24]:
def add_ranks(combination):

    combination = combination.sort_values('Share', ascending=False)
    combination['Rk'] = list(range(1, combination.shape[0]+1))

    combination = combination.sort_values('predictions', ascending=False)
    combination['Rk_predicted'] = list(range(1, combination.shape[0]+1))
    
    combination['Diff'] = combination['Rk'] - combination['Rk_predicted']
    
    return combination

In [25]:
# Example: 0th array -- Year 2006 predictions. look for ranks of top 5
add_ranks(all_predictions[0]).sort_values('Diff', ascending=False)

Player  Share  predictions   Rk  Rk_predicted  Diff
4164   Alonzo Mourning  0.000     0.055599  461            30   431
11154    Zach Randolph  0.000     0.033866  464            53   411
3857   Stephon Marbury  0.000     0.037602  452            44   408
5172         Pau Gasol  0.000     0.078654  395            13   382
3319    Chris Andersen  0.000     0.031875  434            58   376
...                ...    ...          ...  ...           ...   ...
7568      Jerome James  0.000    -0.024128   24           420  -396
6942      Scot Pollard  0.000    -0.027193   30           434  -404
3332        P.J. Brown  0.001    -0.024279   16           422  -406
8187     Jason Collins  0.000    -0.039404   39           460  -421
7566     Danny Fortson  0.000    -0.042899   26           462  -436

[464 rows x 6 columns]

In [26]:
# Example: 0th array -- Year 2006 predictions. look for ranks of top 5
ranking = add_ranks(all_predictions[0]).sort_values('Diff', ascending=False)
ranking[ ranking['Rk'] <= 5 ].sort_values('Diff', ascending=False)

Player  Share  predictions  Rk  Rk_predicted  Diff
8515        Tim Duncan  0.258     0.174906   4             2     2
4175  Shaquille O'Neal  0.813     0.293396   2             1     1
2813     Dirk Nowitzki  0.275     0.112860   3             7    -4
916      Allen Iverson  0.189     0.075288   5            18   -13
4388        Steve Nash  0.839     0.051237   1            33   -32

We'll make the function for `backtesting` for easier applications ahead.

In [27]:
years = list(range(2000, 2024))

def backtesting(stats, model, year, predictors):

    avg_precisions = []
    all_predictions = []

    ## We need 5 years of data for predictions:
    for year in years[5:] :
        train = stats[ stats['Year']  < year]
        test  = stats[ stats['Year'] == year]

        model.fit(train[predictors], train['Share'])

        predictions = model.predict(test[predictors])
        predictions = pd.DataFrame(predictions, columns=['predictions'],
                                                  index=test.index)
        combination = pd.concat([test[['Player', 'Share']], predictions], axis=1)
        
        combination = add_ranks(combination)

        all_predictions.append(combination)
        avg_precisions.append(find_avg_precision(combination))
        
    mean_avg_precision = sum(avg_precisions) / len(avg_precisions)
    all_predictions    = pd.concat(all_predictions)
    
    return mean_avg_precision, avg_precisions, all_predictions

In [28]:
mean_ap, aps, all_predictions = backtesting(stats, reg, years[5:], predictors)

In [29]:
mean_ap

0.6925775627091418

### Diagnosing The Model's Performance

We've done backtesting, putting in the ranks for relevant parameters, now let's look at how the model performs. We'll look for correlations in each predictors to know which ones are the main MVP predictor of the model.

In [30]:
# Make a DataFrame of correlations
corr_ridge = pd.concat([pd.Series(predictors), pd.Series(reg.coef_)], axis=1)
# Sort the values in descending order
corr_ridge.sort_values(0, ascending=False)

0         1
13  eFG%  0.088189
26  Year -0.000549
29  W/L%  0.035754
27     W  0.000081
19   TRB -0.029761
23   TOV -0.004790
21   STL  0.008287
33   SRS -0.000557
25   PTS  0.008739
31  PS/G -0.000618
24    PF -0.003788
32  PA/G  0.000070
17   ORB  0.025472
3     MP -0.003600
28     L -0.000325
2     GS  0.000008
30    GB  0.000293
1      G  0.000076
15   FTA  0.011384
16   FT% -0.003850
14    FT -0.009032
5    FGA -0.005703
6    FG% -0.150059
4     FG  0.006797
18   DRB  0.036468
22   BLK  0.007929
0    Age  0.000046
20   AST  0.007579
8    3PA  0.000899
9    3P% -0.010831
7     3P -0.008619
11   2PA -0.007494
12   2P%  0.007457
10    2P  0.008682

### Adding More Predictors

We have a `Ridge` accuracy of 69.2577% in predicting MVPs over the years. To improve the performance, we can try adding another predictor. We'll do several experiments, and we'll start taking a ratio of individual player's stat compared to its average in a given season.

In [31]:
stats

Player Pos  Age   Tm   G  GS    MP   FG   FGA    FG%  ...  \
0          A.C. Green  PF   36  LAL  82  82  23.5  2.1   4.7  0.447  ...   
1          Brian Shaw  SG   33  LAL  74   2  16.9  1.7   4.4  0.382  ...   
2        Derek Fisher  PG   25  LAL  78  22  23.1  2.1   6.2  0.346  ...   
3       Devean George  SF   22  LAL  49   1   7.0  1.1   2.9  0.389  ...   
4           Glen Rice  SF   32  LAL  80  80  31.6  5.3  12.3  0.430  ...   
...               ...  ..  ...  ...  ..  ..   ...  ...   ...    ...  ...   
11514   Spencer Hawes  PF   28  MIL  54   1  14.8  2.5   5.1  0.484  ...   
11515     Steve Novak  PF   33  MIL   8   0   2.8  0.3   0.9  0.286  ...   
11516  Terrence Jones  PF   25  MIL  54  12  23.5  4.3   9.1  0.470  ...   
11517      Thon Maker   C   19  MIL  57  34   9.9  1.5   3.2  0.459  ...   
11518      Tony Snell  SG   25  MIL  80  80  29.2  3.1   6.8  0.455  ...   

       Pts Max  Share                Team   W   L   W/L%   GB   PS/G   PA/G  \
0          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
1          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
2          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
3          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
4          0.0    0.0  Los Angeles Lakers  67  15  0.817  0.0  100.8   92.3   
...        ...    ...                 ...  ..  ..    ...  ...    ...    ...   
11514      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   
11515      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   
11516      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   
11517      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   
11518      0.0    0.0     Milwaukee Bucks  42  40  0.512  9.0  103.6  103.8   

        SRS  
0      8.41  
1      8.41  
2      8.41  
3      8.41  
4      8.41  
...     ...  
11514 -0.45  
11515 -0.45  
11516 -0.45  
11517 -0.45  
11518 -0.45  

[11519 rows x 41 columns]

In [32]:
stat_ratios = (stats[['PTS', 'AST', 'STL', 'BLK', '3P', 'Year']]
                   .groupby('Year', as_index=False)
                   .apply(lambda x: x / x.mean())
              ).reset_index()

In [38]:
stat_ratios

PTS_T     AST_R     STL_R     BLK_R      3P_R
0      0.627089  0.540640  0.894701  0.467519  0.000000
1      0.514213  1.459729  0.745584  0.467519  0.512850
2      0.790132  1.513793  1.491168  0.000000  1.794977
3      0.401337  0.108128  0.298234  0.233759  0.769276
4      1.994143  1.189409  0.894701  0.467519  2.820678
...         ...       ...       ...       ...       ...
11514  0.811285  0.435028  1.476263  4.707424  0.000000
11515  0.811285  0.483365  0.492088  0.523047  1.312418
11516  1.238854  1.015066  1.148205  1.046094  2.221015
11517  2.148808  1.353421  0.656117  0.523047  1.716239
11518  0.109633  0.000000  0.000000  0.000000  0.302866

[11519 rows x 5 columns]

In [33]:
stat_ratios.columns

Index(['level_0', 'level_1', 'PTS', 'AST', 'STL', 'BLK', '3P', 'Year'], dtype='object')

In [34]:
del stat_ratios['level_0']
del stat_ratios['level_1']
del stat_ratios['Year']

In [35]:
stat_ratios.columns = ['PTS_T', 'AST_R', 'STL_R', 'BLK_R', '3P_R']
stat_ratios.columns

Index(['PTS_T', 'AST_R', 'STL_R', 'BLK_R', '3P_R'], dtype='object')

In [36]:
stats = pd.concat([stats, stat_ratios], axis=1)

In [37]:
stats

Player Pos  Age   Tm   G  GS    MP   FG   FGA    FG%  ...  \
0          A.C. Green  PF   36  LAL  82  82  23.5  2.1   4.7  0.447  ...   
1          Brian Shaw  SG   33  LAL  74   2  16.9  1.7   4.4  0.382  ...   
2        Derek Fisher  PG   25  LAL  78  22  23.1  2.1   6.2  0.346  ...   
3       Devean George  SF   22  LAL  49   1   7.0  1.1   2.9  0.389  ...   
4           Glen Rice  SF   32  LAL  80  80  31.6  5.3  12.3  0.430  ...   
...               ...  ..  ...  ...  ..  ..   ...  ...   ...    ...  ...   
11514   Spencer Hawes  PF   28  MIL  54   1  14.8  2.5   5.1  0.484  ...   
11515     Steve Novak  PF   33  MIL   8   0   2.8  0.3   0.9  0.286  ...   
11516  Terrence Jones  PF   25  MIL  54  12  23.5  4.3   9.1  0.470  ...   
11517      Thon Maker   C   19  MIL  57  34   9.9  1.5   3.2  0.459  ...   
11518      Tony Snell  SG   25  MIL  80  80  29.2  3.1   6.8  0.455  ...   

        W/L%   GB   PS/G   PA/G   SRS     PTS_T     AST_R     STL_R     BLK_R  \
0      0.817  0.0  100.8   92.3  8.41  0.627089  0.540640  0.894701  0.467519   
1      0.817  0.0  100.8   92.3  8.41  0.514213  1.459729  0.745584  0.467519   
2      0.817  0.0  100.8   92.3  8.41  0.790132  1.513793  1.491168  0.000000   
3      0.817  0.0  100.8   92.3  8.41  0.401337  0.108128  0.298234  0.233759   
4      0.817  0.0  100.8   92.3  8.41  1.994143  1.189409  0.894701  0.467519   
...      ...  ...    ...    ...   ...       ...       ...       ...       ...   
11514  0.512  9.0  103.6  103.8 -0.45  0.811285  0.435028  1.476263  4.707424   
11515  0.512  9.0  103.6  103.8 -0.45  0.811285  0.483365  0.492088  0.523047   
11516  0.512  9.0  103.6  103.8 -0.45  1.238854  1.015066  1.148205  1.046094   
11517  0.512  9.0  103.6  103.8 -0.45  2.148808  1.353421  0.656117  0.523047   
11518  0.512  9.0  103.6  103.8 -0.45  0.109633  0.000000  0.000000  0.000000   

           3P_R  
0      0.000000  
1      0.512850  
2      1.794977  
3      0.769276  
4      2.820678  
...         ...  
11514  0.000000  
11515  1.312418  
11516  2.221015  
11517  1.716239  
11518  0.302866  

[11519 rows x 46 columns]

In [40]:
predictors += ['PTS_T', 'AST_R', 'STL_R', 'BLK_R', '3P_R']

mean_ap, aps, all_predictions = backtesting(stats, reg, years[5:], predictors)

In [41]:
mean_ap

0.6907908114487062

Apparently the accuracy didn't change much from `69.2577%` to `69.079%`.

### Categorical Values as Predictors

We can try adding another predictors by modifying non-numerical values in the `stats` dataset. We see that there are types of player's positions `Pos` and player's teams `Tm`. They're categorical so we can modify it to match the data types for our model.

In [43]:
print(stats['Pos'].unique())

# Convert the data into numeric yet categorical value
stats['NPos'] = stats['Pos'].astype('category').cat.codes

['PF' 'SG' 'PG' 'SF' 'C' 'PG-SG' 'PF-SF' 'SG-PG' 'PF-C' 'SG-SF' 'SF-PF'
 'SF-SG' 'C-PF' 'SG-PF' 'PG-SF' 'SG-PG-SF' 'SF-C']


In [44]:
print(stats['Tm'].unique())

# Convert the data into numeric yet categorical value
stats['NTm'] = stats['Tm'].astype('category').cat.codes

['LAL' 'MIA' 'CLE' 'MIL' 'CHI' 'GSW' 'DAL' 'IND' 'WAS' 'MIN' 'PHO' 'ATL'
 'HOU' 'DEN' 'ORL' 'NOH' 'TOR' 'SAC' 'CHO' 'PHI' 'BOS' 'OKC' 'NJN' 'NOK'
 'LAC' 'UTA' 'CHA' 'MEM' 'SEA' 'NYK' 'NOP' 'POR' 'BRK' 'DET' 'SAS' 'CHH'
 'VAN']


In [45]:
stats

Player Pos  Age   Tm   G  GS    MP   FG   FGA    FG%  ...  \
0          A.C. Green  PF   36  LAL  82  82  23.5  2.1   4.7  0.447  ...   
1          Brian Shaw  SG   33  LAL  74   2  16.9  1.7   4.4  0.382  ...   
2        Derek Fisher  PG   25  LAL  78  22  23.1  2.1   6.2  0.346  ...   
3       Devean George  SF   22  LAL  49   1   7.0  1.1   2.9  0.389  ...   
4           Glen Rice  SF   32  LAL  80  80  31.6  5.3  12.3  0.430  ...   
...               ...  ..  ...  ...  ..  ..   ...  ...   ...    ...  ...   
11514   Spencer Hawes  PF   28  MIL  54   1  14.8  2.5   5.1  0.484  ...   
11515     Steve Novak  PF   33  MIL   8   0   2.8  0.3   0.9  0.286  ...   
11516  Terrence Jones  PF   25  MIL  54  12  23.5  4.3   9.1  0.470  ...   
11517      Thon Maker   C   19  MIL  57  34   9.9  1.5   3.2  0.459  ...   
11518      Tony Snell  SG   25  MIL  80  80  29.2  3.1   6.8  0.455  ...   

        PS/G   PA/G   SRS     PTS_T     AST_R     STL_R     BLK_R      3P_R  \
0      100.8   92.3  8.41  0.627089  0.540640  0.894701  0.467519  0.000000   
1      100.8   92.3  8.41  0.514213  1.459729  0.745584  0.467519  0.512850   
2      100.8   92.3  8.41  0.790132  1.513793  1.491168  0.000000  1.794977   
3      100.8   92.3  8.41  0.401337  0.108128  0.298234  0.233759  0.769276   
4      100.8   92.3  8.41  1.994143  1.189409  0.894701  0.467519  2.820678   
...      ...    ...   ...       ...       ...       ...       ...       ...   
11514  103.6  103.8 -0.45  0.811285  0.435028  1.476263  4.707424  0.000000   
11515  103.6  103.8 -0.45  0.811285  0.483365  0.492088  0.523047  1.312418   
11516  103.6  103.8 -0.45  1.238854  1.015066  1.148205  1.046094  2.221015   
11517  103.6  103.8 -0.45  2.148808  1.353421  0.656117  0.523047  1.716239   
11518  103.6  103.8 -0.45  0.109633  0.000000  0.000000  0.000000  0.302866   

       NPos  NTm  
0         2   15  
1        12   15  
2         5   15  
3         8   15  
4         8   15  
...     ...  ...  
11514     2   18  
11515     2   18  
11516     2   18  
11517     0   18  
11518    12   18  

[11519 rows x 48 columns]

We've changed the categorical values to numerical values, ranging from `1-32`. That being said, this will hurt the linear regression since the number will be calculated into the operation instead of categorized, yet the numbers in these values do not represent the value of team.

We can try implementing another model to resolve this by using `RandomForest()`.

### (2) [`RandomForest()`](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html): Random Forest

`RandomForest()` is an ensemble method, derivative of `DecisionTree()`, where the model branches a decision based on a threshold of a certain parameter it finds best. 

The difference with regular `DecisionTree()` is that `RandomForest()` will have samples of random observations from the original dataset. 

This can create diverse tree models, and therefore reaching a powerful combination that decreases the tendency to overfit, a problem that emerges in `DecisionTree()`.

In [46]:
years[18]

2018

In [47]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(
    n_estimators=50, random_state=1, min_samples_split=5)

# The fitting will take quite a while, 
# so we'll limit the year to 2018 for now.
(mean_ap_rf, aps_rf, 
 all_predictions_rf) = backtesting(stats, rf, years[18:], predictors)

# Adding Ridge Regression for comparison
(mean_ap_reg, aps_reg, 
  all_predictions_reg) = backtesting(stats, reg, years[18:], predictors)

In [48]:
mean_ap_rf

0.700173188044705

In [49]:
mean_ap_reg

0.6907908114487062

We can see that average precision of `RandomForest()` is higher with `70.017%` compared to `Ridge()` regression with `69.079%`.

Still this could've been better. The performance in MVP predictions can be tweaked further by these steps:

* Going with another model (machine learning or deep learning),
* generating more predictors on the set, or
* Adding another bunch years of data through web scraping for more training,

We're going to add another one via another model called `GradientBoosting()`.

### (3) `GradientBoosting()`: Gradient Boosting

`GradientBoosting()` is another ensemble of `DecisionTree()`. What makes `GradientBoosting()` distinct is that this model builds an ensemble of trees one-by-one, and then the residual are calculated and reconstructed to boost the prediction performance.

In [50]:
from sklearn.ensemble import GradientBoostingRegressor

gb = GradientBoostingRegressor(
        n_estimators=50, random_state=1, min_samples_split=5)

# The fitting will take quite a while, 
# so we'll limit the year to 2018 for now.
(mean_ap_gb, aps_gb, 
 all_predictions_gb) = backtesting(stats, gb, years[18:], predictors)

In [51]:
mean_ap_gb

0.7146230169983633

We can see that `GradientBoosting` has the highest average precision so far with `71.462%`

## Model Performance: Average Precision -- `Ridge()` vs `RandomForest()` vs `GradientBoosting()`

Last but not least, let's see the performance of backtesting for all models we've tried. This will take quite a while to run, but we'll have the whole reliable evaluation metrics in the end.

In [52]:
# Adding Ridge Regression for comparison
(mean_ap_reg, aps_reg, 
  all_predictions_reg) = backtesting(stats, reg, years[5:], predictors)

(mean_ap_rf, aps_rf, 
 all_predictions_rf) = backtesting(stats, rf, years[5:], predictors)

(mean_ap_gb, aps_gb, 
 all_predictions_gb) = backtesting(stats, gb, years[5:], predictors)

Training set average precision -- Ridge: 0.6907908114487062
Training set average precision -- Random Forest: 0.700173188044705
Training set average precision -- Gradient Boosting: 0.7146230169983633


In [55]:
print(f'Avg precision, prediction of MVP (2005-2023) - Ridge: {mean_ap_reg}')
print(f'Avg precision, prediction of MVP (2005-2023) - Random Forest: {mean_ap_rf}')
print(f'Avg precision, prediction of MVP (2005-2023) - Gradient Boosting: {mean_ap_gb}')

Avg precision, prediction of MVP (2005-2023) - Ridge: 0.6907908114487062
Avg precision, prediction of MVP (2005-2023) - Random Forest: 0.700173188044705
Avg precision, prediction of MVP (2005-2023) - Gradient Boosting: 0.7146230169983633


# Summary

In this project, we've predicted the Most Valuable Players in NBA from 2000-2023 in a website of https://www.basketball-reference.com . We've collected the `mvps` on each year, `players`' stats, and also the `teams` they're in, combining them all, and clean the data for MVP prediction. 

We've implemented three models of machine learning:

* `Ridge()`, 
* `RandomForest()`, and 
* `GradientBoosting()`. 

Out of the three models we've tried, we've come to a conclusion that `GradientBoosting()` perform the best in predicting Most Valuable Players (MVPs) with an accuracy of average precision of  `71.4623%`.